README.md
## Image Optimization

Images are usually the heaviest components of a web page, both in terms of bytes and number of HTTP requests. Optimizing images on your website is critical to improve your users’ experience, reduce delivery costs and enhance your position in search engine ranking. For example, Google’s Largest Contentful Paint metric in their search ranking algorithm is highly impacted by how much you optimize the images on your website. In the solution, we provide you with a simple and performant solution for image optimization using serverless components such as Amazon CloudFront, Amazon S3 and AWS Lambda.

The proposed architecture is suitable for most common use cases. Image transformation is executed centrally in an AWS Region, only when the image hasn’t been already transformed and stored. The available transformations include resizing and formatting, but can be extended to more operations if needed. Both transformations can be requested by the front-end, with the possibility of automatic format selection done on server side. The architecture is based on S3 for storage, CloudFront for content delivery, and Lambda for image processing. The request flow is explained in the next diagram:

<img src="architecture.png" width="900">

1. The user sends a HTTP request for an image with specific transformations, such as encoding and size. The transformations are encoded in the URL, more precisely as query parameters. An example URL would look like this: https://exmaples.com/images/cats/mycat.jpg?format=webp&width=200.
2. The request is processed by a nearby CloudFront edge location providing the best performance. Before passing the request upstream, a CloudFront Function is executed on viewer request event to rewrite the request URL. CloudFront Functions is a feature of CloudFront that allows you to write lightweight functions in JavaScript for high-scale, latency-sensitive CDN customizations. In our architecture, we rewrite the URL to validate the requested transformations and normalize the URL by ordering transformations and convert them to lower case to increase the cache hit ratio. When an automatic transformation is requested, the function also decides about the best one to apply. For example, if the user asks for the most optimized image format (JPEG,WebP, or AVIF) using the directive format=auto, CloudFront Function will select the best format based on the Accept header present in the request.
3. If the requested image is already cached in CloudFront then there will be a cache hit and the image is returned from CloudFront cache. To increase the cache hit ratio, we enable Origin shield, a feature of CloudFront that acts as an additional layer of caching before the origin, to further offload it from requests. If the Image is not in CloudFront cache, then the request will be forwarded to an S3 bucket, which is created to store the transformed images. If the requested image is already transformed and stored in S3, then it is simply served and cached in CloudFront.
4. Otherwise, S3 will respond with a 403 error code, which is detected by CloudFront’s Origin Failover. Thanks to this native feature, CloudFront retries the same URL but this time using the secondary origin based on Lambda function URL. When invoked, the Lambda function downloads the original image from another S3 bucket, where original images are stored, transforms it using Sharp library, stores the transformed image in S3, then serve it through CloudFront where it will be cached for future requests.

Note the following:

* The transformed image is stored in S3 with a lifecycle policy that deletes it after a certain duration (default of 90 days) to reduce the storage cost. Ideally, you’d set this value according to the duration after which the number of requests to a new image drops significantly. They are created with the same key as the original image in addition to a suffix based on the normalized image transformations. For example, the transformed image in response to /mycat.jpg?format=auto&width=200 would be stored with the key /mycat.jpg/format=webp,width=200 if the automatically detected format was webp. To remove all generated variants of the same image in S3, delete all files listed under the key of the original image /mycat.jpg/*. Transformed images are added to S3 with a Cache-Control header of 1 year. If you need to invalidate all cached variants of an image in CloudFront, use the following invalidation pattern: /mycat.jpg*.
* To prevent from unauthorized invocations of the Lambda function, CloudFront is configured to send a secret key in a Custom origin header, which is validated in the Lambda function before processing the image.

## Deploy the solution using CDK
AWS CDK is an open-source software development framework used to define cloud infrastructure in code and provision it through AWS CloudFormation. Follow these steps in your command line to deploy the image optimization solution with CDK, using the region and account information configured in your AWS CLI.

```
git clone https://github.com/aws-samples/image-optimization.git 
cd image-optimization
npm install
cdk bootstrap
npm run build
cdk deploy

# for sotogubeng:
cdk deploy --context RECORD_NAME=image --context ZONE_NAME=sotogubeng.com --context ZONE_ID=Z0038615LKFBRBB1OMPT --context S3_IMAGE_BUCKET_NAME=image-sotogubeng-com --all --require-approval never
# where [RECORD_NAME].[ZONE_NAME] is the custom domain for the cloudfront distribution. This will also add RECORD_NAME record to hosted zone in Route 53.
```

When the deployment is completed within minutes, the CDK output will include the domain name of the CloudFront distribution created for image optimization (ImageDeliveryDomain =YOURDISTRIBUTION.cloudfront.net). The stack will include an S3 bucket with sample images (OriginalImagesS3Bucket = YourS3BucketWithOriginalImagesGeneratedName). To verify that it is working properly, test the following optimized image URL https:// YOURDISTRIBUTION.cloudfront.net/images/rio/1.jpeg?format=auto&width=300.

Note that when deploying in production, it’s recommended to use an existing S3 bucket where your images are stored. To do that, deploy the stack in the same region of your S3 bucket, using the following parameter: cdk deploy -c S3_IMAGE_BUCKET_NAME=’YOUR_S3_BUCKET_NAME’. The solution allows you to configure other parameters such as whether you want to store transformed images in S3 (STORE_TRANSFORMED_IMAGES), the duration after which transformed images are automatically removed from S3 (STORE_TRANSFORMED_IMAGES), and the Cache-Control header used with transformed images (S3_TRANSFORMED_IMAGE_CACHE_TTL).

## Clean up resources

To remove cloud resources created for this solution, just execute the following command:

```
cdk destroy
```

## License

This library is licensed under the MIT-0 License. See the LICENSE file.

CHANGELOG.md
# Release Notes

## [v1.0.3](https://github.com/IDN-Media/image-optimization.popbela.com/compare/release/1.0.2...IDN-Media:image-optimization.popbela.com:release/1.0.3) - 2023-07-12

### Added
- feat: GNDL-1608 response 404 when image doenst exist in https://github.com/IDN-Media/image-optimization.popbela.com/commit/dff6883a8d9594da2afdff931fe9ff45d355e1aa
### Fix
- fix: enable dns cdn in https://github.com/IDN-Media/image-optimization.popbela.com/commit/5fd8eca7bda55302c54b71a176cd00442f85ee79
- fix: doenst show 404 in https://github.com/IDN-Media/image-optimization.popbela.com/commit/e985515c44cfa535606ad074994fffb024411b95
- fix: adjust response in https://github.com/IDN-Media/image-optimization.popbela.com/commit/1c266bb93790cd10ec7983b55b4e2478991aafa9
- fix: adjust respone 404 in https://github.com/IDN-Media/image-optimization.popbela.com/commit/2c6d350ecc4e41301592fb3d48d517f30d9d7517

## [v1.0.2](https://github.com/IDN-Media/image-optimization.popbela.com/compare/release/1.0.1...IDN-Media:image-optimization.popbela.com:release/1.0.2) - 2023-07-07
### Fix
- Ci: disable dns cdn in https://github.com/IDN-Media/image-optimization.popbela.com/commit/43deec54476f148f52832b2354f92bf104e065f8


## [v1.0.1](https://github.com/IDN-Media/image-optimization.popbela.com/tree/release/1.0.1) - 2023-07-07
### Added
- Feat: Response 200 when access root url in https://github.com/IDN-Media/image-optimization.popbela.com/pull/1
- Feat: body return json in https://github.com/IDN-Media/image-optimization.popbela.com/commit/362e57581688d947c9e87d50d80e41f46d25be13


## Note :
```
- Added = Fitur Baru
- Changed = Perubahan Fitur
- Fixed = Fix Bug
```

functions/image-processing/index.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

const AWS = require('aws-sdk');
const https = require('https');
const Sharp = require('sharp');

const S3 = new AWS.S3({signatureVersion: 'v4',httpOptions: {agent: new https.Agent({keepAlive: true})}}); 
const S3_ORIGINAL_IMAGE_BUCKET = process.env.originalImageBucketName; 
const S3_TRANSFORMED_IMAGE_BUCKET = process.env.transformedImageBucketName; 
const TRANSFORMED_IMAGE_CACHE_TTL = process.env.transformedImageCacheTTL;
const SECRET_KEY = process.env.secretKey;
const LOG_TIMING = process.env.logTiming;

exports.handler = async (event) => {
    // First validate if the request is coming from CloudFront
    if (!event.headers['x-origin-secret-header'] || !(event.headers['x-origin-secret-header'] === SECRET_KEY)) return sendError(403, 'Request unauthorized', event);
    // Validate if this is a GET request
    if (!event.requestContext || !event.requestContext.http || !(event.requestContext.http.method === 'GET')) return sendError(400, 'Only GET method is supported', event);
    // An example of expected path is /rio/images/1.jpg/format=auto,width=100 or /rio/images/1.jpg/original where /rio/images/1.jpg is the path of the original image
    var imagePathArray= event.requestContext.http.path.split('/');
    // get the requested image operations
    var operationsPrefix = imagePathArray.pop(); 
    // get the original image path images/rio/1.jpg
    imagePathArray.shift(); 
    var originalImagePath = imagePathArray.join('/');
    // response 200 when accessing root url
    if (originalImagePath.length == 0) {
        return {
            statusCode: 200,
            body: JSON.stringify({
                "status": "OK"
            }),
        };
    }
    // timing variable
    var timingLog = "perf ";
    var startTime = performance.now();
    // Downloading original image
    let originalImage;
    let contentType;
    try {
        originalImage = await S3.getObject({ Bucket: S3_ORIGINAL_IMAGE_BUCKET, Key: originalImagePath }).promise();
        contentType = originalImage.ContentType;
    } catch (error) {
        return sendError(404, '404 Image Not Found', error);
    }
    let sharpObject = Sharp(originalImage.Body);
    let transformedImage;
    //  execute the requested operations 
    var operationsJSON = {};
    var operationsArray = operationsPrefix.split(',');
    operationsArray.forEach(operation => {
        var operationKV = operation.split("=");
        operationsJSON[operationKV[0]] = operationKV[1];
    });
    timingLog = timingLog + parseInt(performance.now()-startTime) + ' ';
    startTime = performance.now();
    try {
        // check if resizing is requested
        var resizingOptions = {};
        if (operationsJSON['width']) resizingOptions.width = parseInt(operationsJSON['width']);
        if (operationsJSON['height']) resizingOptions.height = parseInt(operationsJSON['height']);
        if (resizingOptions) transformedImage = sharpObject.resize(resizingOptions);
        // check if formatting is requested
        if (operationsJSON['format']) {
            var isLossy = false;
            switch (operationsJSON['format'])
            {
               case 'jpeg': contentType = 'image/jpeg'; isLossy = true; break;
               case 'svg': contentType = 'image/svg+xml'; break;
               case 'gif': contentType = 'image/gif'; break;
               case 'webp': contentType = 'image/webp'; isLossy = true; break;
               case 'png': contentType = 'image/png'; break;
               case 'avif': contentType = 'image/avif'; isLossy = true; break;
               default : contentType = 'image/jpeg'; isLossy = true;
            }
            if (operationsJSON['quality'] && isLossy) {
                transformedImage = transformedImage.toFormat(operationsJSON['format'], {
                    quality: parseInt(operationsJSON['quality']),
                });
            } else transformedImage = transformedImage.toFormat(operationsJSON['format']);
        }
        transformedImage = await transformedImage.toBuffer();
    } catch (error) {
        return sendError(500, 'error transforming image', error);
    }
    timingLog = timingLog + parseInt(performance.now()-startTime) + ' ';
    startTime = performance.now();
    // upload transformed image back to S3 if required in the architecture
    if (S3_TRANSFORMED_IMAGE_BUCKET) {
        try { 
            await S3.putObject({
                Body: transformedImage, 
                Bucket: S3_TRANSFORMED_IMAGE_BUCKET, 
                Key:  originalImagePath + '/' + operationsPrefix, 
                ContentType: contentType,
                Metadata: {
                    'cache-control': TRANSFORMED_IMAGE_CACHE_TTL,
                },
            }, function(err, data) {}).promise();
        } catch (error) {
            sendError('APPLICATION ERROR', 'Could not upload transformed image to S3', error);
        }
    }
    timingLog = timingLog + parseInt(performance.now()-startTime) + ' ';
    if (LOG_TIMING === 'true') console.log(timingLog);
    // return transformed image
    return {
        statusCode: 200,
        body: transformedImage.toString('base64'),
        isBase64Encoded: true,
        headers: {
            'Content-Type': contentType, 
            'Cache-Control': TRANSFORMED_IMAGE_CACHE_TTL 
        }
    };
};

function sendError(code, message, error){
    console.log('APPLICATION ERROR', message);
    console.log(error);
    return {
        statusCode: code,
        body: message,
    };
}


lib/image-optimization-stack.ts
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import { Stack, StackProps, RemovalPolicy, aws_s3 as s3, aws_s3_deployment as s3deploy, aws_cloudfront as cloudfront, aws_cloudfront_origins as origins, aws_lambda as lambda, aws_iam as iam, Duration, CfnOutput, aws_logs as logs} from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { MyCustomResource } from './my-custom-resource';
import { createHash } from 'crypto';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';
import { Certificate } from 'aws-cdk-lib/aws-certificatemanager';
import { ARecord, HostedZone, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';

// Stack Parameters

// related to architecture. If set to false, transformed images are not stored in S3, and all image requests land on Lambda
var STORE_TRANSFORMED_IMAGES = 'true';
// Parameters of S3 bucket where original images are stored
var S3_IMAGE_BUCKET_NAME:string;
// CloudFront parameters
var CLOUDFRONT_CORS_ENABLED = 'true';
// Parameters of transformed images
var S3_TRANSFORMED_IMAGE_EXPIRATION_DURATION = '90'; 
var S3_TRANSFORMED_IMAGE_CACHE_TTL = 'max-age=31622400';
// Lambda Parameters
var LAMBDA_MEMORY = '1500';
var LAMBDA_TIMEOUT = '60';
var LOG_TIMING = 'false';

var ZONE_NAME: string;
var RECORD_NAME: string;
var ZONE_ID: string;

type ImageDeliveryCacheBehaviorConfig = {
  origin: any;
  viewerProtocolPolicy: any;
  cachePolicy: any;
  functionAssociations: any;
  responseHeadersPolicy?:any;
};

type LambdaEnv = {
  originalImageBucketName: string,
  transformedImageBucketName?:any;
  transformedImageCacheTTL: string,
  secretKey: string,
  logTiming: string,
}

export class ImageOptimizationStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // Change stack parameters based on provided context
    STORE_TRANSFORMED_IMAGES = this.node.tryGetContext('STORE_TRANSFORMED_IMAGES') || STORE_TRANSFORMED_IMAGES;
    S3_TRANSFORMED_IMAGE_EXPIRATION_DURATION = this.node.tryGetContext('S3_TRANSFORMED_IMAGE_EXPIRATION_DURATION') || S3_TRANSFORMED_IMAGE_EXPIRATION_DURATION; 
    S3_TRANSFORMED_IMAGE_CACHE_TTL = this.node.tryGetContext('S3_TRANSFORMED_IMAGE_CACHE_TTL') || S3_TRANSFORMED_IMAGE_CACHE_TTL;
    S3_IMAGE_BUCKET_NAME = this.node.tryGetContext('S3_IMAGE_BUCKET_NAME') || S3_IMAGE_BUCKET_NAME;
    CLOUDFRONT_CORS_ENABLED = this.node.tryGetContext('CLOUDFRONT_CORS_ENABLED') || CLOUDFRONT_CORS_ENABLED;
    LAMBDA_MEMORY = this.node.tryGetContext('LAMBDA_MEMORY') || LAMBDA_MEMORY;
    LAMBDA_TIMEOUT = this.node.tryGetContext('LAMBDA_TIMEOUT') || LAMBDA_TIMEOUT;
    LOG_TIMING = this.node.tryGetContext('LOG_TIMING') || LOG_TIMING;
    RECORD_NAME = this.node.tryGetContext('RECORD_NAME') || RECORD_NAME;
    ZONE_NAME = this.node.tryGetContext('ZONE_NAME') || ZONE_NAME;
    ZONE_ID = this.node.tryGetContext('ZONE_ID') || ZONE_ID;

    const domainName = `${RECORD_NAME}.${ZONE_NAME}`;

    // Create secret key to be used between CloudFront and Lambda URL for access control
    const SECRET_KEY = createHash('md5').update(this.node.addr).digest('hex') ;

    // For the bucket having original images, either use an external one, or create one with some samples photos.
    var originalImageBucket;
    var transformedImageBucket;
    var sampleWebsiteDelivery;

    if (S3_IMAGE_BUCKET_NAME) {
      originalImageBucket = s3.Bucket.fromBucketName(this,'imported-original-image-bucket', S3_IMAGE_BUCKET_NAME);
      new CfnOutput(this, 'OriginalImagesS3Bucket', {
        description: 'S3 bucket where original images are stored',
        value: originalImageBucket.bucketName
      });  
    } else {
      originalImageBucket = new s3.Bucket(this, 's3-sample-original-image-bucket', {
        removalPolicy: RemovalPolicy.DESTROY,
        blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
        encryption: s3.BucketEncryption.S3_MANAGED,
        enforceSSL: true,
        autoDeleteObjects: true, 
      });
      new s3deploy.BucketDeployment(this, 'DeployWebsite', {
        sources: [s3deploy.Source.asset('./image-sample')],
        destinationBucket: originalImageBucket,
        destinationKeyPrefix: 'images/rio/',
      });
      var sampleWebsiteBucket = new s3.Bucket(this, 's3-sample-website-bucket', {
        removalPolicy: RemovalPolicy.DESTROY,
        blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
        encryption: s3.BucketEncryption.S3_MANAGED,
        enforceSSL: true,
        autoDeleteObjects: true, 
      });

      sampleWebsiteDelivery = new cloudfront.Distribution(this, 'websiteDeliveryDistribution', {
        comment: 'image optimization - sample website',
        defaultRootObject: 'index.html',
        defaultBehavior: {
          origin: new origins.S3Origin(sampleWebsiteBucket),
          viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
        }
      });
      new CfnOutput(this, 'SampleWebsiteDomain', {
        description: 'Sample website domain',
        value: sampleWebsiteDelivery.distributionDomainName
      });
      new CfnOutput(this, 'SampleWebsiteS3Bucket', {
        description: 'S3 bucket use by the sample website',
        value: sampleWebsiteBucket.bucketName
      });  
      new CfnOutput(this, 'OriginalImagesS3Bucket', {
        description: 'S3 bucket where original images are stored',
        value: originalImageBucket.bucketName
      });  
    }
    
    // create bucket for transformed images if enabled in the architecture
    if (STORE_TRANSFORMED_IMAGES === 'true') {
      transformedImageBucket = new s3.Bucket(this, 's3-transformed-image-bucket', {
        removalPolicy: RemovalPolicy.DESTROY,
        autoDeleteObjects: true, 
        lifecycleRules: [
            {
              expiration: Duration.days(parseInt(S3_TRANSFORMED_IMAGE_EXPIRATION_DURATION)),
            },
          ],
      });
    }

    // prepare env variable for Lambda 
    var lambdaEnv: LambdaEnv = {
      originalImageBucketName: originalImageBucket.bucketName,
      transformedImageCacheTTL: S3_TRANSFORMED_IMAGE_CACHE_TTL,
      secretKey: SECRET_KEY,
      logTiming: LOG_TIMING,
    };
    if (transformedImageBucket) lambdaEnv.transformedImageBucketName = transformedImageBucket.bucketName;

    // IAM policy to read from the S3 bucket containing the original images
    const s3ReadOriginalImagesPolicy = new iam.PolicyStatement({
      actions: ['s3:GetObject'],
      resources: ['arn:aws:s3:::'+originalImageBucket.bucketName+'/*'],
    });
    
    const cloudfrontOAI = new cloudfront.OriginAccessIdentity(
      this, 'CloudFrontOriginAccessIdentity'
    );
    originalImageBucket.addToResourcePolicy(new iam.PolicyStatement({
        actions: ['s3:GetObject'],
        resources: ['arn:aws:s3:::'+originalImageBucket.bucketName+'/*'],
        principals: [new iam.CanonicalUserPrincipal(
            cloudfrontOAI.cloudFrontOriginAccessIdentityS3CanonicalUserId)],
    }));

    // statements of the IAM policy to attach to Lambda
    var iamPolicyStatements = [s3ReadOriginalImagesPolicy];

    // Create Lambda for image processing
    var lambdaProps = {
      runtime: lambda.Runtime.NODEJS_16_X, 
      code: lambda.Code.fromAsset('functions/image-processing'),
      handler: 'index.handler',
      timeout: Duration.seconds(parseInt(LAMBDA_TIMEOUT)),
      memorySize: parseInt(LAMBDA_MEMORY),
      environment: lambdaEnv,
      logRetention: logs.RetentionDays.ONE_DAY,
    };
    var imageProcessing = new lambda.Function(this, 'image-optimization', lambdaProps);

    // Enable Lambda URL
    const imageProcessingURL = imageProcessing.addFunctionUrl({
      authType: lambda.FunctionUrlAuthType.NONE,
    });

    // Leverage a custom resource to get the hostname of the LambdaURL
    const imageProcessingHelper = new MyCustomResource(this, 'customResource', {
      Url: imageProcessingURL.url
    });

    // Create a CloudFront origin: S3 with fallback to Lambda when image needs to be transformed, otherwise with Lambda as sole origin
    var imageOrigin;

    if (transformedImageBucket) {
      imageOrigin = new origins.OriginGroup ({
        primaryOrigin: new origins.S3Origin(transformedImageBucket, { }),
        fallbackOrigin: new origins.HttpOrigin(imageProcessingHelper.hostname, {
          customHeaders: {
            'x-origin-secret-header': SECRET_KEY,
          },
        }), 
        fallbackStatusCodes: [403],
      });

      // write policy for Lambda on the s3 bucket for transformed images
      var s3WriteTransformedImagesPolicy = new iam.PolicyStatement({
        actions: ['s3:PutObject'],
        resources: ['arn:aws:s3:::'+transformedImageBucket.bucketName+'/*'],
      });
      iamPolicyStatements.push(s3WriteTransformedImagesPolicy);
    } else {
      console.log("else transformedImageBucket");
      imageOrigin = new origins.HttpOrigin(imageProcessingHelper.hostname, {
        customHeaders: {
          'x-origin-secret-header': SECRET_KEY,
        },
      });
    }

    // attach iam policy to the role assumed by Lambda
    imageProcessing.role?.attachInlinePolicy(
      new iam.Policy(this, 'read-write-bucket-policy', {
        statements: iamPolicyStatements,
      }),
    );

    // Create a CloudFront Function for url rewrites
    const urlRewriteFunction = new cloudfront.Function(this, 'urlRewrite', {
      code: cloudfront.FunctionCode.fromFile({filePath: 'functions/url-rewrite/index.js',}),
      functionName: `urlRewriteFunction${this.node.addr}`, 
    });

    const cachePolicy = new cloudfront.CachePolicy(this, `ImageCachePolicy${this.node.addr}`, {
      defaultTtl: Duration.hours(24),
      maxTtl: Duration.days(365),
      minTtl: Duration.seconds(30),
      queryStringBehavior: cloudfront.CacheQueryStringBehavior.all()
    });

    var imageDeliveryCacheBehaviorConfig:ImageDeliveryCacheBehaviorConfig  = {
      origin: imageOrigin,
      viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
      cachePolicy,
      functionAssociations: [{
        eventType: cloudfront.FunctionEventType.VIEWER_REQUEST,
        function: urlRewriteFunction,
      }],
    }

    if (CLOUDFRONT_CORS_ENABLED === 'true') {
      // Creating a custom response headers policy. CORS allowed for all origins.
      const imageResponseHeadersPolicy = new cloudfront.ResponseHeadersPolicy(this, `ResponseHeadersPolicy${this.node.addr}`, {
        responseHeadersPolicyName: 'ImageResponsePolicy',
        corsBehavior: {
          accessControlAllowCredentials: false,
          accessControlAllowHeaders: ['*'],
          accessControlAllowMethods: ['GET'],
          accessControlAllowOrigins: ['*'],
          accessControlMaxAge: Duration.seconds(600),
          originOverride: false,
        },
        // recognizing image requests that were processed by this solution
        customHeadersBehavior: {
          customHeaders: [
            { header: 'x-aws-image-optimization', value: 'v1.0', override: true },
            { header: 'vary', value: 'accept', override: true },
          ],
        }
      });  
      imageDeliveryCacheBehaviorConfig.responseHeadersPolicy = imageResponseHeadersPolicy;
    }

    const certificateUsEastArn = StringParameter.fromStringParameterName(this, 'certificateUsEastArn', '/idn/popbela/certificateUsEastArn').stringValue;
    const certificateUsEast = Certificate.fromCertificateArn(this, "certificateUsEast", certificateUsEastArn);

    const imageDelivery = new cloudfront.Distribution(this, 'imageDeliveryDistribution', {
      comment: 'image optimization - image delivery',
      defaultBehavior: imageDeliveryCacheBehaviorConfig,
      additionalBehaviors: {
        "/*/*/*.svg":{
          origin: new origins.S3Origin(originalImageBucket, {
            originAccessIdentity: cloudfrontOAI
          }),
          allowedMethods: cloudfront.AllowedMethods.ALLOW_GET_HEAD,
          viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
          cachePolicy,
        }
      },
      // alternate domain image.popbela.com and cdn.popbela.com
      domainNames: [domainName,`cdn.${ZONE_NAME}`],
      certificate: certificateUsEast,
    });

    const hostedZone = HostedZone.fromHostedZoneAttributes(this, 'HostedZone', {
      hostedZoneId: ZONE_ID,
      zoneName: ZONE_NAME,
    });

    new ARecord(this, 'AliasRecord', {
      recordName: RECORD_NAME,
      zone: hostedZone,
      target: RecordTarget.fromAlias(new CloudFrontTarget(imageDelivery)),
    });

    // alias for domain cdn.popbela.com
    // note di production cname manual
    //new ARecord(this, 'CdnAliasRecord', {
    //  recordName: 'cdn',
    //  zone: hostedZone,
    //  target: RecordTarget.fromAlias(new CloudFrontTarget(imageDelivery)),
   // });

    new CfnOutput(this, 'ImageDeliveryDomain', {
      description: 'Domain name of image delivery',
      value: imageDelivery.distributionDomainName
    });
  }
}
